### PR TITLE
GH-2658 Literal/ValueFactory java.time support

### PR DESCRIPTION
**Pull request by knoan**
GitHub issue resolved: #2658

Briefly describe the changes proposed in this PR:

- add Literal.temporalAccessor/AmountValue() methods
- add ValueFactory.createLiteral(TemporalAccessor/Amount) methods
- implement new methods in AbstractLiteral and AbstractValueFactory

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits down to one or a few meaningful commits
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

### CODE REVIEW

**Comment by knoan** [comment]
Mostly done. Comments welcome on tests and on the last commit.

`SimpleValueFactory` now transitively inherits `ValueFactory.createLiteral(TemporalAccessor)` from `oer.model.base.AbstractValueFactory`, so that tests don't break and existing code can be upgraded without migrating to `AbstractValueFactory`; as far as I can see, all other factory methods are overridden, so the existing code base should be safe.

Taking into account that we are adding new functionalities, everything should be ok, but I'd like an additional assessment about returning an `AbstractLiteral` from `SimpleValueFactory`.

**Comment by knoan** [comment]
P.S. Hope this is not late for 3.5.0: if that's not the case, `@since` tags should be updated

File: core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java [hunk 1]
```diff
@@ -175,6 +333,25 @@ public BigDecimal decimalValue() {
 		return value(BigDecimal::new);
 	}
 
+	@Override
+	public TemporalAccessor temporalValue() throws DateTimeException {
+		return value(label -> {
+
+			final TemporalAccessor value = label.startsWith("--") ? DASH_FORMATTER.parse(label)
+					: label.length() >= 8 && label.charAt(2) == ':' ? OFFSET_TIME_FORMATTER.parse(label)
+							: DATETIME_FORMATTER.parse(label);
```
Comment (jeenbroekstra) [inline core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java:342]: Can we not use nested/chained ternaries? YMMV, but I find them hard to read. Might be improved with linebreaks in different places, but I think an if-else is easier on the eye (if a little more verbose):

```java
TemporalAccessor value;
if (label.startsWith("--")) {
    value = DASH_FORMATTER.parse(label);
} else if (label.length() >= 8 && label.charAt(2) == ':' ) {
   value = OFFSET_TIME_FORMATTER.parse(label);
} else {
  value = DATETIME_FORMATTER.parse(label);
}
```

or perhaps more clean:

```java
final TemporalAccessor value = getFormatter(label).parse();

... 

private static DateFormatter getFormatter(String label) {
    if(label.startsWith("--")) {
       return DASH_FORMATTER;
   } 
   if (label.length() >= 8 && label.charAt(2) == ":") {
       return OFFSET_TIME_FORMATTER;     
   }
   return DATETIME_FORMATTER;
}
```

**Review by jeenbroekstra — COMMENTED** [review]
This is definitely not too late for 3.5, we did a milestone build but we made it clear that it was not considered feature-complete, so there room for this. 

I think it looks very solid, thanks again @knoan. I only really have a few code organization comments (see inline).

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/AbstractValueFactory.java [hunk 2]
```diff
@@ -39,7 +39,7 @@
  *             {@link org.eclipse.rdf4j.model.base.AbstractValueFactory}.
  */
 @Deprecated
-public abstract class AbstractValueFactory implements ValueFactory {
+public abstract class AbstractValueFactory extends org.eclipse.rdf4j.model.base.AbstractValueFactory {
```
Comment (jeenbroekstra) [inline core/model/src/main/java/org/eclipse/rdf4j/model/impl/AbstractValueFactory.java:42]: I actually need to look at this, see #2655. I will likely refactor SimpleValueFactory to directly inherit from the new AbstractValueFactory, copying all the old AbstractValueFactory code over and leaving it orphaned.

Comment (knoan) [reply core/model/src/main/java/org/eclipse/rdf4j/model/impl/AbstractValueFactory.java:42]: > I actually need to look at this, see #2655. I will likely refactor SimpleValueFactory to directly inherit from the new AbstractValueFactory, copying all the old AbstractValueFactory code over and leaving it orphaned.

As long as SimpleValueFactory inherits from the new AbstractValueFactor tests won't break.

File: core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java [hunk 3]
```diff
@@ -72,6 +87,149 @@
...
+			.appendLiteral(':')
+			.appendValue(MINUTE_OF_HOUR, 2)
+			.appendLiteral(':')
+			.appendValue(SECOND_OF_MINUTE, 2)
+
+			.optionalStart()
+			.appendFraction(NANO_OF_SECOND, 0, 9, true)
+
+			.toFormatter();
+
```
[the remark this answers is not in the document]
Comment (knoan) [reply core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java]: I refactor to migrated all support fields and parsing code to the relevant specialised nested classes: now we are in the position to easily promote them to top level, if we opt to do so.

For the time being, however, I'd leave them inside AbstractLiteral, in order not to relax encapsulation and symmetry with other generic implementations in other Abstract* value classes.

File: core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java (hunk 1 above)
Comment (knoan) [reply core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java]: Done.

**Comment by knoan** [comment]
Before merging I'd like to add support for XSD duration datatypes: allot me a couple of days.

**Comment by knoan** [comment]
> Before merging I'd like to add support for XSD duration datatypes: allot me a couple of days.

I was able to cram that into an unforeseen free slot, but it was quite nasty ;-P

`java.time` doesn't include `TemporalAmount` implementations able to represent XML schema duration values including both date and time components (e.g. `P1YT1H`): `Period` supports only date-based components and `Duration` supports only time-based ones.

I reckon that's conservatively done in order to prevent average/estimated year/month/day durations to creep into calendrical calculations involving also time-based-components, but that forced us to implement it by ourselves along with the companion parsing/formatting methods: some compatibility corner cases are likely to have done it through tests…

**Comment by knoan** [comment]
P.S. I've already squashed the incremental changes, as they were very fragmented

**Comment by knoan** [comment]
Handled new temporal methods also in delegating ValueFactories.

**Comment by knoan** [comment]
Utilities methods in `Literals` could also be eventually completed, but I'd defer that after the whole class is reviewed for `Simple*` and other suspicious dependencies.

IMHO, most of these utilities could be factored to Optional-based methods on Value/Literal interfaces: hid `AbstractLiteral.value(Function<String, V> mapper)` method to avoid interferences with possible future developments in this direction.

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/AbstractValueFactory.java (hunk 2 above)
Comment (jeenbroekstra) [reply core/model/src/main/java/org/eclipse/rdf4j/model/impl/AbstractValueFactory.java:42]: I've merged my update for #2655. It shouldn't break anything for you but might be worth rebasing, just to make sure.

**Comment by jeenbroekstra** [comment]
> I reckon that's conservatively done in order to prevent average/estimated year/month/day durations to creep into calendrical calculations involving also time-based-components, but that forced us to implement it by ourselves along with the companion parsing/formatting methods: some compatibility corner cases are likely to have done it through the tests…

Wow. Good stuff. So does this mean we are no longer depending on javax.xml.DatatypeFactory to process these kinds of datatype strings? If so that is actually quite a relief, because that thing was ancient and still wasn't fully updated for XSD 1.1.

**Comment by knoan** [comment]
> So does this mean we are no longer depending on javax.xml.DatatypeFactory to process these kinds of datatype strings?

Correct. The relevant code is in the following methods:

- AbstractLiteral.TemporalAccessorLiteral.parseTemporalAccessor(String)
- AbstractLiteral.TemporalAmountLiteral.parseTemporalAmount(String)/toString(TemporalAmount)

There's still room for improvement, but the code is completely hidden, so we are completely at liberty to refactor it as we see fit.

Time permitting, it would also be worth extending parsing/formatting methods, as current codecs were hastily derived from java.time ISO formats, which XSD 2 restricts in a number of ways.

Given that we are dealing with RDF 1.1 literal expected to be well-formed, I'd stick to strict parsing, leaving lenient versions to eventual RIO upgrades.

**Comment by knoan** [comment]
I'm realising that the current code base (which I took as a starting point for this PR) defines calendar values referring to XSD 2:

> A calendar representation can be given for literals whose label conforms to the syntax of the following <a href="http://www.w3.org/TR/xmlschema-2/">XML Schema datatypes</a> …

RDF 1.1 specs on the other hand are still based on XSD 1.1 (https://www.w3.org/TR/rdf11-concepts/#xsd-datatypes).

If we are not in a hurry, I'd review the whole matter to achieve full conformance with RDF 1.1, along the following lines:

- redefine new temporal methods wrt RDF 1.1 / XSD 1.1 (no impact on existing code as we are introducing them anew)
- add parsing/formatting tests based on XSD 1.1 specs

In order to provide better interoperability with XSD 2 (and an easier upgrade path), I'd still ignore temporal datatypes dropped from the latter version of the standard, that is `xsd:dateTimeStamp`, `xsd:yearMonthDuration` and `xsd:dayTimeDuration`.

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/AbstractValueFactory.java (hunk 2 above)
Comment (knoan) [reply core/model/src/main/java/org/eclipse/rdf4j/model/impl/AbstractValueFactory.java:42]: Rebased

File: core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java (hunk 1 above)
Comment (jeenbroekstra) [reply core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractLiteral.java]: Great, thanks!

**Comment by jeenbroekstra** [comment]
> I'm realising that the current code base (which I took as a starting point for this PR) defines calendar values referring to XSD 2:

I haven't looked in detail but I believe that to just be a mistake in the javadoc. AFAIK our actual calendar value support is based on XSD 1.1 specs (and in corner cases even only XSD 1.0).

**Comment by jeenbroekstra** [comment]
> If we are not in a hurry, I'd review the whole matter to achieve full conformance with RDF 1.1, along the following lines:
> 
>  - redefine new temporal methods wrt RDF 1.1 / XSD 1.1 (no impact on existing code as we are introducing them anew)
> - add parsing/formatting tests based on XSD 1.1 specs

Making sure we are at least conformant with RDF 1.1 is of course a good idea. 

> In order to provide better interoperability with XSD 2 (and an easier upgrade path), I'd still ignore temporal datatypes dropped from the latter version of the standard, that is xsd:dateTimeStamp, xsd:yearMonthDuration and xsd:dayTimeDuration.

When you say "ignore", would that simply mean that if someone creates a literal with one of those datatypes (and valid lexical value for that type), we won't support `getTemporalValue` on such a literal? In other words: people are free to create, store, and query such literals, and even use the legacy Date/Calendar-based methods to process it, but the newly added support based on the java.time api will not handle such values. Correct?

**Comment by knoan** [comment]
> When you say "ignore", would that simply mean that if someone creates a literal with one of those datatypes (and valid lexical value for that type), we won't support `getTemporalValue` on such a literal?

Legacy XSD 1.1 datatypes were defined as restriction of the base datatypes retained in XSD 2:

- assigning a more specific value to the base type is still formally correct (and IMHO more aligned with current LOD practice);
- if the literal is explicitly created using a label and one of the legacy datatypes (`ValueFactory.createLiteral(String, IRI)`),  `LIteral.getTemporal*Value()` will return the correct temporal value, as long as the label is well-formed;
- if the literal is created using a temporal value (`ValueFactory.createLiteral(TemporalAccessor/Amount)`), `LIteral.getTemporal*Value()` will return the supplied value, while `Literal.getDatatype()` will return the base XSD 1.1/2 datatype rather than the more specific XSD 1.1 datatype later dropped from XSD 2, i.e.:
    - xsd:dateTimeStamp >> xsd:dateTime
    - xsd:yearMonthDuration >> xsd:duration
    - xsd:dayTimeDuration >> xsd:duration

> In other words: people are free to create, store, and query such literals, and even use the legacy Date/Calendar-based methods to process it, but the newly added support based on the java.time api will not handle such values. Correct?

**Comment by knoan** [comment]
> > I'm realising that the current code base (which I took as a starting point for this PR) defines calendar values referring to XSD 2:
> 
> I haven't looked in detail but I believe that to just be a mistake in the javadoc. AFAIK our actual calendar value support is based on XSD 1.1 specs (and in corner cases even only XSD 1.0). Or were you referring to your new implementation being based on XSD 2?

I was mislead by the reference to XSD 2 in the javadocs and derived new conformance tests from that specs rather than XSD 1.1: I don't really expect any deviation, but I'd rather review the whole matter.

**Comment by knoan** [comment]
I reviewed XSD 1.1/2 conformance, confirming the overall scheme already discusessed in https://github.com/eclipse/rdf4j/pull/2659#issuecomment-729512460:

- `Literal.temporalAccessorValue()` parses lexical values according to XSD 1.1 rules for date/time datatypes 
- `Literal.temporalAmountValue()` parses lexical values according to XSD 2 rules for the `xsd:duration` datatype
- `ValueFactory.createLiteral(TemporalAccessor)` assigns an XSD 1.1 date/time datatype ignoring `xsd:dateTimeStamp`
- `ValueFactory.createLiteral(TemporalAmount)` always assigns the `xsd:duration` datatype ignoring more specific XSD 1.1 duration datatypes later dropped from XSD 2

While not strictly compliant with RDF 1.1, this approach is simpler, way easier to implement and (imho) more aligned with current LOD practice; known deviation are documented in javadocs.

Tests reviewed as well: some corner cases likely to be still slipping through, but probably not worth further immediate effort.

As far as I'm concerned ready for final reviews.

**Review by jeenbroekstra — APPROVED** [review]
All looks very solid to me. Great work @knoan ! 

I think we have to start considering a proper 3.5 release soon, so that we can get this out to the masses...
